File: packages/xbridge-ios/Sources/XBridgeiOS/Security/XBridgeSecurityPolicy.swift
// SPDX-License-Identifier: MIT
//
// XBridgeSecurityPolicy.swift
// XBridgeiOS
//
// Created by XBridge SDK on 2024-01-01.
//

import Foundation

/// A security policy that controls which origins are allowed to use the
/// native bridge.
///
/// This is a **defense-in-depth** layer. The primary security gate lives on
/// the Flutter side (`WebViewBridgePolicy` in `xbridge_flutter`). This native
/// policy can be applied as an additional check when the app intercepts
/// URL loading or evaluates bridge calls.
///
/// Example:
/// ```swift
/// let policy = XBridgeSecurityPolicy.allowlist(["https://app.example.com"])
/// // Later:
/// if policy.allows(origin: "https://app.example.com") {
///     // permit bridge call
/// }
/// ```
public struct XBridgeSecurityPolicy {

    /// The set of allowed origins (e.g. `"https://app.example.com"`).
    public let allowedOrigins: Set<String>

    /// If `true`, all origins are permitted (useful for development).
    public let allowAll: Bool

    /// Private initializer — use the static factory methods.
    private init(allowedOrigins: Set<String>, allowAll: Bool) {
        self.allowedOrigins = allowedOrigins
        self.allowAll = allowAll
    }

    // MARK: - Factory methods

    /// Create a policy that permits all origins.
    /// - Warning: Use only in development environments.
    public static func allowAll() -> XBridgeSecurityPolicy {
        return XBridgeSecurityPolicy(allowedOrigins: [], allowAll: true)
    }

    /// Create a policy that permits only the specified origins.
    public static func allowlist(_ origins: Set<String>) -> XBridgeSecurityPolicy {
        return XBridgeSecurityPolicy(allowedOrigins: origins, allowAll: false)
    }

    // MARK: - Evaluation

    /// Check if the given origin is allowed by this policy.
    ///
    /// - Parameter origin: The origin string (e.g. `"https://app.example.com"`).
    /// - Returns: `true` if the origin is permitted.
    public func allows(origin: String) -> Bool {
        if allowAll {
            return true
        }
        return allowedOrigins.contains(origin)
    }
}

File: packages/xbridge-android/src/main/java/io/xbridge/XBridgeSecurityPolicy.kt
package io.xbridge

/**
 * Security policy for XBridge bridge calls — a defense-in-depth origin
 * allowlist that supplements the Flutter-side [WebViewBridgePolicy].
 *
 * The **primary** security gate lives in the Flutter `BridgeController`,
 * which checks the page origin before dispatching. This native policy is a
 * secondary check for calls that arrive via the sync bypass
 * ([XBridgeSyncInterface]) or directly via the fallback MethodChannel.
 *
 * @property allowedOrigins  Set of allowed origins (e.g.
 *                            `setOf("https://app.example.com")`).
 * @property allowAll         If `true`, all origins are allowed (use for
 *                            development only).
 */
data class XBridgeSecurityPolicy(
    val allowedOrigins: Set<String> = emptySet(),
    val allowAll: Boolean = false,
) {
    companion object {
        /** Deny all origins — secure default. Use allowAll() for development only. */
        fun denyAll(): XBridgeSecurityPolicy = XBridgeSecurityPolicy(allowedOrigins = emptySet(), allowAll = false)

        /** Allow all origins — development convenience. */
        fun allowAll(): XBridgeSecurityPolicy = XBridgeSecurityPolicy(allowAll = true)

        /** Restrict to an explicit allowlist of origins. */
        fun allowlist(origins: Set<String>): XBridgeSecurityPolicy =
            XBridgeSecurityPolicy(allowedOrigins = origins, allowAll = false)

        /** Default: deny all origins except those in the allowlist. */
        fun allowlist(vararg origins: String): XBridgeSecurityPolicy =
            allowlist(origins.toSet())
    }

    /**
     * Returns `true` if [origin] is permitted by this policy.
     * An empty allowlist with `allowAll = false` denies everything.
     *
     * Origin comparison is case-insensitive and ignores trailing slashes
     * and default ports (80 for http, 443 for https) to avoid false
     * rejections from minor URL formatting differences.
     */
    fun allows(origin: String?): Boolean {
        if (allowAll) return true
        if (origin == null) return false
        // Reject "null" origin (sandboxed iframes, data: URIs) and wildcard "*"
        // to match the Rust WS server's security checks.
        if (origin == "null" || origin == "*") return false
        val normalized = normalizeOrigin(origin)
        return allowedOrigins.any { normalizeOrigin(it) == normalized }
    }

    /**
     * Normalize an origin string: lowercase, strip trailing slash, strip
     * default ports (443 for https, 80 for http).
     */
    private fun normalizeOrigin(origin: String): String {
        var o = origin.trim().lowercase()
        // Strip trailing slash
        while (o.endsWith("/")) {
            o = o.dropLast(1)
        }
        // Strip default ports
        o = o.removeSuffix(":443")
        if (o.startsWith("https://")) {
            o = o.removePrefix("https://")
            o = "https://${o.removeSuffix(":443")}"
        } else if (o.startsWith("http://")) {
            o = o.removePrefix("http://")
            o = "http://${o.removeSuffix(":80")}"
        }
        return o
    }
}
